Sources/SKTestSupport/IndexedSingleSwiftFileWorkspace.swift

```diff
@@ -0,0 +1,125 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+import SKCore
+import Foundation
+import ISDBTibs
+import SourceKitLSP
+import TSCBasic
+
+public struct IndexedSingleSwiftFileWorkspace {
+  enum Error: Swift.Error {
+    case swiftcNotFound
+  }
+
+  public let testClient: TestSourceKitLSPClient
+  public let fileURI: DocumentURI
+  public let positions: DocumentPositions
+
+  public init(
+    _ markedText: String,
+    testName: String = #function
+  ) async throws{
+
+    // Build file paths
+
+    let testBaseName = testName.prefix(while: \.isLetter)
+    let testWorkspaceDirectory = FileManager.default.temporaryDirectory
+      .realpath
+      .appendingPathComponent(String(testBaseName))
+      .appendingPathComponent(UUID().uuidString)
+
+    try FileManager.default.createDirectory(at: testWorkspaceDirectory, withIntermediateDirectories: true)
+
+    let testFileURL = testWorkspaceDirectory.appendingPathComponent("test.swift")
+    let indexURL = testWorkspaceDirectory.appendingPathComponent("index")
+    let indexDBURL = testWorkspaceDirectory.appendingPathComponent("index-db")
+    let toolchain = ToolchainRegistry.shared.default!
+    guard let swiftc = toolchain.swiftc?.asURL else {
+      throw Error.swiftcNotFound
+    }
+
+    // Create workspace with source file and compile_commands.json
+
+    try extractMarkers(markedText).textWithoutMarkers.write(to: testFileURL, atomically: false, encoding: .utf8)
+
+    let compilerArguments: [String] = [
+      testFileURL.path,
+      "-sdk", TibsBuilder.defaultSDKPath!,
+      "-index-store-path", indexURL.path,
+      "-index-ignore-system-modules",
+      "-o", testWorkspaceDirectory.appendingPathComponent("test.out").path
+    ]
+
+    let compilationDatabase = JSONCompilationDatabase(
+      [
+        JSONCompilationDatabase.Command(
+          directory: testWorkspaceDirectory.path,
+          filename: testFileURL.path,
+          commandLine: [swiftc.path] + compilerArguments
+        )
+      ]
+    )
+    let encoder = JSONEncoder()
+    encoder.outputFormatting = .prettyPrinted
+    try encoder.encode(compilationDatabase).write(to: testWorkspaceDirectory.appendingPathComponent("compile_commands.json"))
+
+    // Run swiftc to build the index store
+    try ProcessRunner(executableURL: swiftc, arguments: compilerArguments)
```

Comment:
My thoughts were: Running a full compile shouldn’t be noticeably slower in these reduced test cases than just type checking and it’s closer to what a user might do + easier to understand because fewer `swiftc` arguments. That’s what I chose to just do a full compile here. Happy to change it if you disagree, though.

Reply:
I was going to say that `-typecheck` is closer to what we actually do, but I suppose that's not true for LSP since we only have index while building. So doesn't really bother me either way.

Reply:
Let’s keep it with building for now.